CTFd/api/v1/challenges.py

```diff
@@ -123,53 +124,54 @@ def get(self, query_args):
 
         # This can return None (unauth) if visibility is set to public
         user = get_current_user()
+        # Admins get a shortcut to see all challenges despite pre-requisites
+        admin_view = is_admin() and request.args.get("view") == "admin"
 
-        # Admins can request to see everything
-        if is_admin() and request.args.get("view") == "admin":
-            challenges = (
-                Challenges.query.filter_by(**query_args)
-                .filter(*filters)
-                .order_by(Challenges.value)
-                .all()
-            )
-            solve_ids = set([challenge.id for challenge in challenges])
+        # We only set a condition for matching user solves if there is a user
+        # and they have an account ID (user mode or in a team in teams mode)
+        if user is not None and user.account_id is not None:
+            user_solved_cond = (Solves.account_id == user.account_id)
         else:
-            challenges = (
-                Challenges.query.filter(
-                    and_(Challenges.state != "hidden", Challenges.state != "locked")
-                )
-                .filter_by(**query_args)
-                .filter(*filters)
-                .order_by(Challenges.value)
-                .all()
-            )
-
-            if user:
-                solve_ids = (
-                    Solves.query.with_entities(Solves.challenge_id)
-                    .filter_by(account_id=user.account_id)
-                    .order_by(Solves.challenge_id.asc())
-                    .all()
-                )
-                solve_ids = set([value for value, in solve_ids])
-
-                # TODO: Convert this into a re-useable decorator
-                if is_admin():
-                    pass
-                else:
-                    if config.is_teams_mode() and get_current_team() is None:
-                        abort(403)
-            else:
-                solve_ids = set()
+            user_solved_cond = false()
+        # This query counts the number of solves per challenge, as well as the
+        # sum of correct solves made by the current user per the condition
+        # above (which should probably only be 0 or 1!)
+        solves_q = db.session.query(
```

Comment:
This doesn't take into account any of the extra logic around hidden or banned solvers which the `/challenges/<id>` endpoint does :( Also I probably need to respect `score_visible() and accounts_visible()`, and freeze time for solve timestamps.